src/Text/Pandoc/Writers/ConTeXt.hs

```diff
@@ -252,33 +252,42 @@ blockToConTeXt HorizontalRule = return $ "\\thinrule" <> blankline
 -- If this is ever executed, provide a default for the reference identifier.
 blockToConTeXt (Header level attr lst) = sectionHeader attr level lst
 blockToConTeXt (Table caption aligns widths heads rows) = do
-    let colDescriptor colWidth alignment = (case alignment of
-                                               AlignLeft    -> 'l'
-                                               AlignRight   -> 'r'
-                                               AlignCenter  -> 'c'
-                                               AlignDefault -> 'l'):
-           if colWidth == 0
-              then "|"
-              else ("p(" ++ printf "%.2f" colWidth ++ "\\textwidth)|")
-    let colDescriptors = "|" ++ concat (
-                                 zipWith colDescriptor widths aligns)
     headers <- if all null heads
                   then return empty
-                  else liftM ($$ "\\HL") $ tableRowToConTeXt heads
+                  else liftM ("\\startxtablehead[head]" $$) $
+                       liftM ($$ "\\stopxtablehead") $ tableRowToConTeXt aligns widths heads
     captionText <- inlineListToConTeXt caption
-    rows' <- mapM tableRowToConTeXt rows
-    return $ "\\placetable" <> (if null caption
-                                   then brackets "none"
-                                   else empty)
-                            <> braces captionText $$
-             "\\starttable" <> brackets (text colDescriptors) $$
-             "\\HL" $$ headers $$
-             vcat rows' $$ "\\HL" $$ "\\stoptable" <> blankline
+    rows' <- mapM (tableRowToConTeXt aligns widths) rows
+    return $ "\\startplacetable" <> brackets (
+      (if null caption
+        then "location=none"
+        else ("caption=" <> braces captionText))
+      ) $$
+      "\\startxtable" $$ headers $$
+      (if null rows
+        then ""
+        else "\\startxtablebody[body]" $$ vcat (init rows') $$ "\\stopxtablebody" $$
+             "\\startxtablefoot[foot]" $$ (last rows') $$ "\\stopxtablefoot") $$
+      "\\stopxtable" $$ "\\stopplacetable" <> blankline
 
-tableRowToConTeXt :: PandocMonad m => [[Block]] -> WM m Doc
-tableRowToConTeXt cols = do
-  cols' <- mapM blockListToConTeXt cols
-  return $ vcat (map ("\\NC " <>) cols') $$ "\\NC\\AR"
+tableRowToConTeXt :: PandocMonad m => [Alignment] -> [Double] -> [[Block]] -> WM m Doc
+tableRowToConTeXt aligns widths cols = do
+  cells <- mapM tableCellToConTeXt $ zip3 aligns widths cols
+  return $ ("\\startxrow") $$ vcat cells $$ "\\stopxrow"
+
+tableCellToConTeXt :: PandocMonad m => (Alignment, Double, [Block]) -> WM m Doc
+tableCellToConTeXt (align, width, blocks) = do
+  cellContents <- blockListToConTeXt blocks
+  let colWidth = if width == 0
+        then ""
+        else ",width=" <> braces (text (printf "%.2f\\textwidth" width))
```

Comment:
This is ugly Haskell code. A better option is to use something like (sorry, it has been ages since I last wrote Haskell)
```
let colWidth = if width = 0 
    then Nothing
    else  Just $ "width=" <> braces ( ... )
```
Then on line 289, one could use

```
.... brackets (intercalate ", " [halign, fromMaybe "" colWidth])  <>
```

Reply:
Sorry, this is the first bit of Haskell I wrote *ever*.  Your suggestion doesn't work (as is)
```
src/Text/Pandoc/Writers/ConTeXt.hs:285:32:
    Could not deduce (Data.String.IsString [a0])
      arising from the literal ‘"align=right"’
    from the context (PandocMonad m)
      bound by the type signature for
                 tableCellToConTeXt :: PandocMonad m =>
                                       (Alignment, Double, [Block]) -> WM m Doc
      at src/Text/Pandoc/Writers/ConTeXt.hs:278:23-79
    The type variable ‘a0’ is ambiguous
    Relevant bindings include
      halign :: [a0] (bound at src/Text/Pandoc/Writers/ConTeXt.hs:284:7)
    Note: there is a potential instance available:
      instance Data.String.IsString [Char] -- Defined in ‘Data.String’
    In the expression: "align=right"
    In a case alternative: AlignLeft -> "align=right"
    In the expression:
      case align of {
        AlignLeft -> "align=right"
        AlignRight -> "align=left"
        AlignCenter -> "align=middle"
        AlignDefault -> "align=right" }

src/Text/Pandoc/Writers/ConTeXt.hs:290:16:
    Couldn't match expected type ‘Doc’ with actual type ‘[a0]’
    Relevant bindings include
      halign :: [a0] (bound at src/Text/Pandoc/Writers/ConTeXt.hs:284:7)
    In the first argument of ‘brackets’, namely
      ‘(intercalate ", " [halign, fromMaybe "" colWidth])’
    In the first argument of ‘(<>)’, namely
      ‘brackets (intercalate ", " [halign, fromMaybe "" colWidth])’

src/Text/Pandoc/Writers/ConTeXt.hs:290:55:
    Couldn't match type ‘Doc’ with ‘[a0]’
    Expected type: Maybe [a0]
      Actual type: Maybe Doc
    Relevant bindings include
      halign :: [a0] (bound at src/Text/Pandoc/Writers/ConTeXt.hs:284:7)
    In the second argument of ‘fromMaybe’, namely ‘colWidth’
    In the expression: fromMaybe "" colWidth
```

Reply:
`intercalate` doesn't work on elements of type `Doc` because it is essentially `concat $ intersperse` and `concat` is not defined for `Doc`.  One has to use `hcat $ intersperse` instead.